### PR TITLE
Fix ntuple detector selection algorithm

### PR DESCRIPTION
**Pull request by alessandro-Doyensec**
## Context

While adding acceptance tests to `telegrambotapitoken.NewDetector`, I encountered a known issue in the `ntuple.Detector` logic. The previous best-effort algorithm failed to select the correct non-overlapping tuples, causing `TestDetector_Detect/multiple_keys` to fail.

Example of failing testcase:

```
tgram1://key_which_is_long1 tgram2://key_which_is_long2 tgram3://key_which_is_long3
```

Would return the following tuples:

- `tgram2:key_which_is_long1`
- `tgram3:key_which_is_long2`

## Solution

The "best" tuples selection problem can be translated to the "Weighted Interval Scheduling" which has a known O(N log N) solution.

## Additional changes

- Added a `TestDetectorAcceptance` for the `telegrambotapitoken.NewDetector`
- Added the `ntuple_benchmark_test.go` file to verify the new algorithm performance (see below `BenchmarkDetector`)
- Fixed some testcases were the order of expected data previously incorrect.

## Benchmarks

```sh
➜  osv-scalibr git:(main) ✗ go test github.com/google/osv-scalibr/veles/secrets/common/ntuple -bench=BenchmarkDetector -benchmem -cpuprofile=cpu.prof -memprofile=mem.prof
goos: darwin
goarch: arm64
pkg: github.com/google/osv-scalibr/veles/secrets/common/ntuple
cpu: Apple M1 Pro
BenchmarkDetector/Sparse_100KB-10         	    7801	    151992 ns/op	  144923 B/op	    2169 allocs/op
BenchmarkDetector/Dense_100KB-10          	       3	 473453667 ns/op	780943840 B/op	11826162 allocs/op
BenchmarkDetector/Extreme_1MB-10          	       1	12863143417 ns/op	22574156080 B/op	341974842 allocs/op
PASS
ok  	github.com/google/osv-scalibr/veles/secrets/common/ntuple	15.877s
➜  osv-scalibr git:(fix-ntuple-detector-with-multiple-tuples) go test github.com/google/osv-scalibr/veles/secrets/common/ntuple -bench=BenchmarkDetector -benchmem -cpuprofile=cpu.prof -memprofile=mem.prof
goos: darwin
goarch: arm64
pkg: github.com/google/osv-scalibr/veles/secrets/common/ntuple
cpu: Apple M1 Pro
BenchmarkDetector/Sparse_100KB-10         	   14007	     84578 ns/op	   60316 B/op	     336 allocs/op
BenchmarkDetector/Dense_100KB-10          	      10	 108605642 ns/op	284892340 B/op	 1489976 allocs/op
BenchmarkDetector/Extreme_1MB-10          	       1	2828185625 ns/op	8216542416 B/op	42854505 allocs/op
PASS
ok  	github.com/google/osv-scalibr/veles/secrets/common/ntuple	5.478s
```

---

**Note:**

Currently there is a linting error in a file which wasn't modified in this PR.

### CODE REVIEW

File: veles/secrets/cloudflareapitoken/detector_test.go [hunk 1]
```diff
@@ -174,8 +174,8 @@ func TestDetector_truePositives(t *testing.T) {
 			input: `CLOUDFLARE_API_TOKEN="` + testKey + `"
 CF_API_KEY=` + testKey[:len(testKey)-1] + `a`,
 			want: []veles.Secret{
-				cloudflareapitoken.CloudflareAPIToken{Token: testKey[:len(testKey)-1] + "a"},
```
Comment (alessandro-Doyensec) [inline veles/secrets/cloudflareapitoken/detector_test.go:177]: This change just modify the ordering of the `want` secrets

File: veles/secrets/gcpoauth2client/detector_test.go [hunk 2]
```diff
@@ -223,29 +223,6 @@ second_secret: GOCSPX-DuplicateSecret1234567890123`,
 				},
 			},
 		},
-		{
-			name: "deduplication_test_-_multiple_pairs_with_overlapping_credentials",
```
Comment (alessandro-Doyensec) [inline veles/secrets/gcpoauth2client/detector_test.go:227]: This test-case is no longer supported, one secret (or match) can be used only in a tuple.

File: veles/secrets/salesforceoauth2jwt/detector_test.go [hunk 3]
```diff
@@ -181,13 +181,13 @@ valid.user@example.org
 ` + privateKeySample,
 			want: []veles.Secret{
 				salesforceoauth2jwt.Credentials{
-					ID:         "3MVGXYZ987654321ABCDEFGHIJKLM",
```
Comment (alessandro-Doyensec) [inline veles/secrets/salesforceoauth2jwt/detector_test.go:184]: This change just modified the order of the `want` secrets to the correct one

File: veles/secrets/vapid/detector_test.go [hunk 4]
```diff
@@ -210,6 +210,24 @@ func TestDetector_Detect(t *testing.T) {
 				},
 			},
 		},
+		{
+			name: "very bad example",
```
Comment (alessandro-Doyensec) [inline veles/secrets/vapid/detector_test.go]: "Bad" example which add a "wrong" public key between a valid pair of keys.

Used to test the new implementation of the `ntuple.Detector`

Comment (erikvarga) [reply veles/secrets/vapid/detector_test.go]: Let's use a more descriptive name for the specific edge case we're testing. e.g. something like "wrong_public_key_between_valid_key_pairs" ?

File: veles/secrets/salesforceoauth2client/detector_test.go [hunk 5]
```diff
@@ -170,7 +169,6 @@ yuvrajapp.2.my.salesforce.com`,
 			if err != nil {
 				t.Errorf("Detect() error: %v, want nil", err)
 			}
-			fmt.Printf("got = %+v\n", got)
```
Comment (erikvarga) [inline veles/secrets/salesforceoauth2client/detector_test.go]: Remove debug log

File: veles/secrets/common/ntuple/ntuple_benchmark_test.go [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+	for i := 0; i < size-50; i += density {
+		// Pick a random token from our 3 finders
+		token := tokens[rng.Intn(len(tokens))]
+		copy(buf[i:], token)
+	}
+
+	return buf
+}
+
+// Launch using
```
Comment (erikvarga) [inline veles/secrets/common/ntuple/ntuple_benchmark_test.go]: Let's add this to the Makefile as a "make test_benchmark" command

File: veles/secrets/common/ntuple/ntuple.go [hunk 7]
```diff
@@ -20,219 +20,258 @@
...
+type Detector struct {
+	MaxElementLen uint32
+	MaxDistance   uint32
+	Finders       []Finder
+	FromTuple     func([]Match) (veles.Secret, bool)
+	FromPartial   func(Match) (veles.Secret, bool)
 }
 
-var _ veles.Detector = &Detector{}
+// Detect implements the veles.Detector interface.
```
Comment (erikvarga) [inline veles/secrets/common/ntuple/ntuple.go]: Would be useful to add a summary of the algorithm here similar to the previous comment (but it can be shorter / high-level)

File: veles/secrets/common/ntuple/ntuple.go [hunk 8]
```diff
@@ -20,219 +20,258 @@
...
+	var out []veles.Secret
+	var pos []int
+	for _, t := range selected {
+		out = append(out, secretsFromTuple[t])
+		pos = append(pos, t.Start)
+	}
+	return out, pos
+}
+
+func (d *Detector) collect(b []byte) ([]*Tuple, [][]Match) {
```
Comment (erikvarga) [inline veles/secrets/common/ntuple/ntuple.go:124]: Let's also add a short comment on the private functions of the detector for what exactly the params are and what they return.

File: veles/secrets/vapid/detector_test.go (hunk 4 above)
Comment (alessandro-Doyensec) [reply veles/secrets/vapid/detector_test.go]: yeah makes sense, will do

File: veles/secrets/common/ntuple/ntuple_benchmark_test.go (hunk 6 above)
Comment (alessandro-Doyensec) [reply veles/secrets/common/ntuple/ntuple_benchmark_test.go]: Should I also remove the comment on how to launch it?

Comment (erikvarga) [reply veles/secrets/common/ntuple/ntuple_benchmark_test.go]: You can keep something like "launch using make test_benchmark" here

**Comment by alessandro-Doyensec** [comment]
Hello @erikvarga 

Thanks for the review, every conversation should be resolved now
